camel/models/sglang_model.py

```diff
@@ -0,0 +1,153 @@
+
+from openai import OpenAI, Stream
+
+from camel.configs import SGLANG_API_PARAMS, SGLANGConfig
+from camel.messages import OpenAIMessage
+from camel.models import BaseModelBackend
+from camel.types import (
+    ChatCompletion,
+    ChatCompletionChunk,
+    ModelType,
+)
+from camel.utils import BaseTokenCounter, OpenAITokenCounter
+
+
+class SGLANGModel(BaseModelBackend):
+    r"""SGLANG service interface.
+
+    Args:
+        model_type (Union[ModelType, str]): Model for which a backend is
+            created.
+        model_config_dict (Optional[Dict[str, Any]], optional): A dictionary
+            that will be fed into:obj:`openai.ChatCompletion.create()`. If
+            :obj:`None`, :obj:`SGLANGConfig().as_dict()` will be used.
+            (default: :obj:`None`)
+        api_key (Optional[str], optional): The API key for authenticating with
+            the model service. SGLANG doesn't need API key, it would be ignored
+            if set. (default: :obj:`None`)
+        url (Optional[str], optional): The url to the model service. If not
+            provided, :obj:`"http://localhost:8000/v1"` will be used.
+            (default: :obj:`None`)
+        token_counter (Optional[BaseTokenCounter], optional): Token counter to
+            use for the model. If not provided, :obj:`OpenAITokenCounter(
+            ModelType.GPT_4O_MINI)` will be used.
+            (default: :obj:`None`)
+
+    Reference: https://sgl-project.github.io/backend/openai_api_completions.html
+    """
+
+    def __init__(
+        self,
+        model_type: Union[ModelType, str],
+        model_config_dict: Optional[Dict[str, Any]] = None,
+        api_key: Optional[str] = None,
+        url: Optional[str] = None,
+        token_counter: Optional[BaseTokenCounter] = None,
+    ) -> None:
+        if model_config_dict is None:
+            model_config_dict = SGLANGConfig().as_dict()
+        url = url or os.environ.get("SGLANG_BASE_URL")
+        super().__init__(
+            model_type, model_config_dict, api_key, url, token_counter
+        )
+        if not self._url:
+            self._start_server()
+        # Use OpenAI cilent as interface call sglang
+        self._client = OpenAI(
+            timeout=60,
+            max_retries=3,
+            api_key="Set-but-ignored",  # required but ignored
+            base_url=self._url,
+        )
+
+    def _start_server(self) -> None:
+        r"""Starts the sglang server in a subprocess."""
+        try:
+            subprocess.Popen(
+                ["sglang", "server", "--port", "30000"],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            self._url = "http://localhost:30000/v1"
+            print(
+                f"Ollama server started on {self._url} "
+                f"for {self.model_type} model."
+            )
+        except Exception as e:
+            print(f"Failed to start sglang server: {e}.")
```

Comment:
use SGLang native feature as we discussed

Reply:
@Wendong-Fan @liuxukun2000 

Hello. This is Chenyang from SGLang team. Regarding the launch SGLang engine, you can try these two ways:

```python
popen_launch_server(
            cls.model, cls.base_url, timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH
        )
```

This is from https://github.com/sgl-project/sglang/blob/main/test/srt/test_update_weights_from_disk.py

And, alternatively, you can try this:

```python
from sglang.utils import (
    execute_shell_command,
    wait_for_server,
    terminate_process,
    print_highlight,
)

server_process = execute_shell_command(
    "python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --port 30000 --host 0.0.0.0"
)

wait_for_server("http://localhost:30000")
```

This is from our docs:

https://sgl-project.github.io/backend/openai_api_completions.html